connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -200,283 +229,149 @@ private Long ensureLeaderConfig(long maxOffset, WorkerCoordinator coordinator) {
-            Map<String, ConnectorsAndTasks> toExplicitlyRevoke =
-                    performTaskRevocation(activeAssignments, currentWorkerAssignment);
-
-            log.debug("Connector and task to revoke assignments: {}", toRevoke);
-
-            toExplicitlyRevoke.forEach(
-                (worker, assignment) -> {
-                    ConnectorsAndTasks existing = toRevoke.computeIfAbsent(
-                        worker,
-                        v -> new ConnectorsAndTasks.Builder().build());
-                    existing.connectors().addAll(assignment.connectors());
-                    existing.tasks().addAll(assignment.tasks());
-                }
-            );
-            canRevoke = toExplicitlyRevoke.size() == 0;
+            Map<String, ConnectorsAndTasks> loadBalancingRevocations =
+                    performLoadBalancingRevocations(configured, nextWorkerAssignment);
+
+            log.trace("Load-balancing revocations for each worker: {}", loadBalancingRevocations);
+            addAll(toRevoke, loadBalancingRevocations);
+            canRevoke = ConnectorsAndTasks.combine(loadBalancingRevocations.values()).isEmpty();
         } else {
             canRevoke = delay == 0;
         }
 
-        assignConnectors(completeWorkerAssignment, newSubmissions.connectors());
-        assignTasks(completeWorkerAssignment, newSubmissions.tasks());
-        log.debug("Current complete assignments: {}", currentWorkerAssignment);
-        log.debug("New complete assignments: {}", completeWorkerAssignment);
+        final ConnectorsAndTasks toAssign = ConnectorsAndTasks.combine(created, lostAssignmentsToReassign.build());
+        assignConnectors(nextWorkerAssignment, toAssign.connectors());
+        assignTasks(nextWorkerAssignment, toAssign.tasks());
 
-        Map<String, Collection<String>> currentConnectorAssignments =
-                currentWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::connectors));
-        Map<String, Collection<ConnectorTaskId>> currentTaskAssignments =
-                currentWorkerAssignment.stream().collect(Collectors.toMap(WorkerLoad::worker, WorkerLoad::tasks));
-        Map<String, Collection<String>> incrementalConnectorAssignments =
-                diff(connectorAssignments, currentConnectorAssignments);
-        Map<String, Collection<ConnectorTaskId>> incrementalTaskAssignments =
-                diff(taskAssignments, currentTaskAssignments);
+        // The complete set of connectors and tasks that will be running on each worker after this rebalance
+        final Map<String, ConnectorsAndTasks> nextAssignments = assignments(nextWorkerAssignment);
 
-        log.debug("Incremental connector assignments: {}", incrementalConnectorAssignments);
-        log.debug("Incremental task assignments: {}", incrementalTaskAssignments);
+        log.debug("Current complete assignments: {}", memberAssignments);
+        log.debug("New complete assignments: {}", nextAssignments);
```

Comment:
Just minor suggestion and can be ignored. What about changing `New` to `Next`? Because using variable `nextAssignments` for parameter, so logging `Next` looks more clear. 

```suggestion
        log.debug("New complete assignments: {}", nextAssignments);
```

Reply:
That's fair! I was copying the wording in the [existing log message](https://github.com/apache/kafka/blob/4ad439c56db551697ea366dc4986eee5fc288706/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java#L290). Probably best to keep the updated log message consistent with the new variable name like you suggest--another way to go about that is to rename the variable from `nextAssignments` to `newAssignments`. Any preference between the two?

Reply:
@C0urante 
Thank you for the reply:)

Maybe keeping `next` is looking better for me. Because if we change `new` to `next`, we need to change all variable names and consider huge side effects.

Reply:
Fair enough! Done.